Add service for computing the optimal number of shards for data streams

andreidan opened the pull request:
This adds the `DataStreamAutoShardingService` that will compute the optimal number of shards for a data stream and return a recommendation as to when to apply it (a time interval we call cool down which is 0 when the auto sharding recommendation can be applied immediately).

This also introduces a `DataStreamAutoShardingEvent` object that will be stored in the data stream metadata to indicate the last auto sharding event that was applied to a data stream and its cluster state representation looks like so:
```
"auto_sharding": {
 "trigger_index_name": ".ds-logs-nginx-2024.02.12-000002",
 "target_number_of_shards": 3,
 "event_timestamp": 1707739707954
}
```

The auto sharding service is not used in this PR, so the auto sharding event will not be stored in the data stream metadata, but the required infrastructure to configure it is in place.

andreidan committed:
Add service for computing the optimal number of shards for data streams

This adds the `DataStreamAutoShardingService` that will compute the
optimal number of shards for a data stream and return a recommendation
as to when to apply it (a time interval we call cool down which is 0
when the auto sharding recommendation can be applied immediately).

This also introduces a `DataStreamAutoShardingEvent` object that will be
stored in the data stream metadata to indicate the last auto sharding
event that was applied to a data stream and its cluster state
representation looks like so:
```
"auto_sharding": {
 "trigger_generation": 2,
 "trigger_index_name": ".ds-logs-nginx-2024.02.12-000002",
 "target_number_of_shards": 3,
 "event_timestamp": 1707739707954
}
```

The auto sharding service is not used in this PR, so the auto sharding
event will not be stored in the data stream metadata, but the required
infrastructure to configure it is in place.

andreidan committed:
javadoc and compilation

andreidan committed:
fix logging

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

andreidan commented:
@dakrone here's a draft PR illustrating the usage of the service

andreidan committed:
NO_CHANGE_REQUIRED events report cool down 0

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+            .toList();
+
+        // assume the current write index load is the highest observed and look back to find the actual maximum
+        double maxIndexLoadWithinCoolingPeriod = writeIndexLoad;
+        for (IndexWriteLoad writeLoad : writeLoadsWithinCoolingPeriod) {
+            double totalIndexLoad = 0;
+            for (int shardId = 0; shardId < writeLoad.numberOfShards(); shardId++) {
+                final OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    totalIndexLoad += writeLoadForShard.getAsDouble();
+                }
+            }
+
+            if (totalIndexLoad > maxIndexLoadWithinCoolingPeriod) {
+                maxIndexLoadWithinCoolingPeriod = totalIndexLoad;
+            }
+        }
+        return maxIndexLoadWithinCoolingPeriod;
```
I don't know whether this is any cleaner :) but I played some code golf:

```suggestion
        return Math.max(
            writeIndexLoad,
            writeLoadsWithinCoolingPeriod.stream()
                .mapToDouble(
                    load -> IntStream.range(0, load.numberOfShards())
                        .mapToObj(load::getWriteLoadForShard)
                        .mapToDouble(optLoad -> optLoad.orElse(0))
                        .sum()
                )
                .max()
                .orElse(0)
        );
```

andreidan replied:
If you don't have a strong opinion here I'd rather keep a nested loop (as in PR) compare to streaming within a stream and mapping within a map :) 
Happy to apply the suggestion if you think it's more readable.

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+            double totalIndexLoad = 0;
+            for (int shardId = 0; shardId < writeLoad.numberOfShards(); shardId++) {
+                final OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    totalIndexLoad += writeLoadForShard.getAsDouble();
+                }
```
This can be 
```suggestion
                totalIndexLoad += writeLoadForShard.orElse(0);
```

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+        return innerCalculate(metadata, dataStream, writeIndexLoad, nowSupplier);
+    }
+
+    private AutoShardingResult innerCalculate(Metadata metadata, DataStream dataStream, double writeIndexLoad, LongSupplier nowSupplier) {
```
It feels like this function `innerCalculate` is really two or three functions, and maybe should be split up that way (since it's fairly large). One for checking for an optimal shard increase, one for checking whether there's still a remaining decrease cooldown, and one for checking for the optimal shard decrease. What do you think?

andreidan replied:
Good shout, done.

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+    // Visible for testing
+    static long computeOptimalNumberOfShards(int minNumberWriteThreads, int maxNumberWriteThreads, double indexingLoad) {
+        return Math.max(
+            Math.min(roundUp(indexingLoad / ((double) minNumberWriteThreads / 2)), 3),
+            roundUp(indexingLoad / ((double) maxNumberWriteThreads / 2))
```
Random comment, we can avoid the `(double)` casting by using decimals in the constants (if you prefer it)
```suggestion
            Math.min(roundUp(indexingLoad / (minNumberWriteThreads / 2.0)), 3),
            roundUp(indexingLoad / (maxNumberWriteThreads / 2.0))
```

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+    }
+
+    private static long roundUp(double value) {
+        return Math.round(Math.ceil(value));
```
We can also do this as
```suggestion
        return (long) Math.ceil(value);
```
and avoid the call to `Math.round`, but really, whichever you prefer is fine, it's the most minor of nits :)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -1013,6 +1040,11 @@ public void writeTo(StreamOutput out) throws IOException {
         PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), INDEX_MODE);
         PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> DataStreamLifecycle.fromXContent(p), LIFECYCLE);
         PARSER.declareBoolean(ConstructingObjectParser.optionalConstructorArg(), ROLLOVER_ON_WRITE_FIELD);
+        PARSER.declareObject(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> DataStreamAutoShardingEvent.fromXContent(p),
+            AUTO_SHARDING_FIELD
+        );
```
Would we want this behind a check for the feature flag, the same way we have a check for the failure store feature flag? If we did, we'll need to adjust the `args[n]` offsets above.

andreidan replied:
We don't have a feature flag, and I wasn't planning on adding one. The setting to enable the functionality (only available in certain environments) should act as the flag to enable/disable the feature. 

Were you thinking we need a feature flag?

dakrone replied:
If we never have a situation where the xcontent pieces get written out to state, then it's fine not to have one. If we do, then we'll need one to ensure we don't have an unrecoverable cluster due to serialization issues.

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+    }
+
+    // Visible for testing
+    static double getMaxIndexLoadWithinCoolingPeriod(
```
I think some javadocs for this may help understanding what it's doing in the future

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -412,6 +421,10 @@ public DataStreamLifecycle getLifecycle() {
         return lifecycle;
     }
 
+    public DataStreamAutoShardingEvent getAutoShardingEvent() {
```
Can we add a small blurb like 
```suggestion
    /**
     * Returns the latest auto sharding event that happened for this data stream
     */
    public DataStreamAutoShardingEvent getAutoShardingEvent() {
```
to clarify it's the *last* one, and not some amalgamation of all of them or something else

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -1169,6 +1208,34 @@ public DataStream getParentDataStream() {
+     * the entire {@param maxIndexAge} interval so one backing index created before the specified age will also
+     * be return.
+     */
+    public static List<Index> getIndicesWithinMaxAgeRange(
```
Minor suggestion to maybe include "backing" in the name, what do you think? Maybe `getBackingIndicesWithinMaxAgeRange`? Or is that too long?

andreidan replied:
Being a method on the `DataStream` I think this is implied already (and we have quite a few methods on the `DataStream` - all of them? - that perform backing indices computations and eschew the `backing` in the name for readability purposes).

dakrone commented on server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:
```diff
@@ -1671,6 +1689,99 @@ public void testXContentSerializationWithRollover() throws IOException {
+            false,
+            false,
+            false,
+            IndexMode.STANDARD
```
Not that I think it'd make any difference, but should we randomize the mode for this test?

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+
+    public static final NodeFeature DATA_STREAM_AUTO_SHARDING_FEATURE = new NodeFeature("data_stream.auto_sharding");
+
+    public static final Setting<List<String>> DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING = Setting.listSetting(
```
We need to add all these settings to `ClusterSettings.BUILT_IN_CLUSTER_SETTINGS` right?

andreidan replied:
We will in a subsequent PR that makes use of this service (the service itself is now not exposed by any plugin) - and it doesn't have to be in this PR as it's only unit tested. 
(this is to facilitate reviewing and avoid tripping the size of this PR :) ) The follow-up usage is sketched in https://github.com/andreidan/elasticsearch/pull/1 (tests are still to be added there)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAutoShardingEvent.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Represents the last auto sharding event that occured for a data stream.
+ */
+public record DataStreamAutoShardingEvent(String triggerIndexName, long generation, int targetNumberOfShards, long timestamp)
```
We probably need a `DataStreamAutoShardingEventTests` class that extends `SimpleDiffableSerializationTestCase` to check that everything is good with serialization of this class

dakrone commented on server/src/test/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingServiceTests.java:
```diff
@@ -0,0 +1,693 @@
+        builtInClusterSettings.add(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_INCREASE_SHARDS_COOLDOWN);
+        builtInClusterSettings.add(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_DECREASE_SHARDS_COOLDOWN);
+        builtInClusterSettings.add(
+            Setting.boolSetting(
+                DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_ENABLED,
+                false,
+                Setting.Property.Dynamic,
+                Setting.Property.NodeScope
+            )
```
I don't think we should be adding this setting in tests, but making use of it in the non-test code. I'd rather we define it in the regular code (and even add these to the built-in settings assuming the feature flag is enabled) so we don't have to add to the list here

andreidan replied:
As stated above, we're doing this so we don't register the setting in open source.

dakrone commented on server/src/test/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingServiceTests.java:
```diff
@@ -0,0 +1,693 @@
+                    .put(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_ENABLED, true)
+                    .putList(
+                        DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey(),
+                        List.of(dataStreamName + "*")
```
Let's make it not the first data stream in the list (just to test things):
```suggestion
                        List.of("foo", dataStreamName + "*")
```

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+        int targetNumberOfShards,
+        TimeValue coolDownRemaining,
+        @Nullable Double writeLoad
+    ) implements Writeable, ToXContentObject {
```
We probably also want to have a test class for this class that extends `AbstractXContentSerializingTestCase`

andreidan replied:
Ah yes, great catch. Due to the complications of this being used in a rollover Condition (the parsing is different when parsed independently vs via RolloverInfo - I've cried next to the parsing code for a full day to realize that an independent Condition being a named writeable will start parsing _inside_ the object from the first field (`field1` below) when part of `RolloverInfo` e.g. `"auto_sharding": { "field1" ... }` vs outside the object when parsed independently  - the entire parsing code will be part of the rollover condition and not rely necessarily on this object encapsulation. So removing the Writeable/XContent stuff in this PR as will be different in the subsequent PR where it's actually used)

dakrone reviewed (changes requested):
I left some comments, thanks for working on this Andrei!

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+        LongSupplier nowSupplier
+    ) {
+        this.clusterService = clusterService;
+        this.isAutoShardingEnabled = settings.getAsBoolean(DATA_STREAMS_AUTO_SHARDING_ENABLED, false);
```
Yeah, but we are not registering `DATA_STREAM_AUTO_SHARDING_ENABLED_SETTING` in open source as we don't want users to accidentally enable it (even if we leave it undocumented). 
This setting will only be registered and available in other environments. Similar to https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamLifecycle.java#L63

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/GetDataStreamAction.java:
```diff
@@ -348,6 +351,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params, @Nulla
+                    DataStreamAutoShardingEvent autoShardingEvent = dataStream.getAutoShardingEvent();
+                    builder.startObject(AUTO_SHARDING_FIELD.getPreferredName());
+                    autoShardingEvent.toXContent(builder, params);
+                    builder.field(
```
The value being a `TimeValue` it already is a human readable value. I'll make the event timestamp (which is a raw `long` millis in the auto sharding event a human readable field though.

dakrone replied:
It is, but if we do

```java
builder.humanReadableField(
    "time_since_auto_shard_event_millis",
    TIME_SINCE_LAST_AUTO_SHARDING.getPreferredName(),
    autoShardingEvent.getTimeSinceLastAutoShardingEvent(System::currentTimeMillis)
);
```

Then when doing `GET /_data_stream/foo?human` it'll return something like:

```
…
"time_since_last_auto_shard_event": "2.1d",
"time_since_last_auto_shard_event_millis": 129874289
…
```

automatically

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+    private volatile List<String> dataStreamExcludePatterns;
+
+    public enum AutoShardingType {
+        INCREASE_NUMBER_OF_SHARDS,
+        DECREASES_NUMBER_OF_SHARDS,
```
Ah yeah, good catch - that's a typo.

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,473 @@
+
+        @Override
+        public String toString() {
+            return "{ type: "
+                + type
+                + ", currentNumberOfShards: "
+                + currentNumberOfShards
+                + ", targetNumberOfShards: "
+                + targetNumberOfShards
+                + ", coolDownRemaining: "
+                + coolDownRemaining
+                + ", writeLoad: "
+                + writeLoad
+                + " }";
```
I'll work on something different in the subsequent PR (this will be removed). I added this because a rollover `Condition` uses `toString` as the representation for the `RolloverResponse` and cannot be overridden https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/indices/rollover/Condition.java#L68 

Will do something different in the next PR that handles the rollover.

andreidan committed:
Separate classes, renamings, tests

andreidan committed:
Update server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Update server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Update server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Update server/src/test/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingServiceTests.java

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan committed:
Some more renamings

andreidan committed:
javadoc

andreidan committed:
Spilt innerCalculate a bit more

andreidan committed:
line length

andreidan commented:
Thanks for the review @dakrone Think I've addressed most concerns.

andreidan committed:
randomise index mode

andreidan committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan commented:
@elasticmachine test this please

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan committed:
getDecreaseShardsResult is non-nullable

it didn't need to be and it can report the max
write load it encountered in the cooling period
when it recommends NO_CHANGE.

andreidan committed:
Merge branch 'main' into data-stream-autosharding-service

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/GetDataStreamAction.java:
```diff
@@ -179,6 +181,7 @@ public static class DataStreamInfo implements SimpleDiffable<DataStreamInfo>, To
             public static final ParseField TEMPORAL_RANGES = new ParseField("temporal_ranges");
             public static final ParseField TEMPORAL_RANGE_START = new ParseField("start");
             public static final ParseField TEMPORAL_RANGE_END = new ParseField("end");
+            public static final ParseField TIME_SINCE_LAST_AUTO_SHARDING = new ParseField("time_since_last_auto_sharding");
```
Super minor nit, but how do you feel about using `time_since_last_auto_shard_event` or `time_since_last_auto_sharding_event` to avoid making "sharding" into a verb?

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/AutoShardingResult.java:
```diff
@@ -0,0 +1,37 @@
+ * period that needs to lapse before the current recommendation should be applied.
+ * <p>
+ * If auto sharding is not applicable for a data stream (e.g. due to
+ * {@link DataStreamAutoShardingService#DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING}) the target number of shards will be 0 and cool down
```
```suggestion
 * {@link DataStreamAutoShardingService#DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING}) the target number of shards will be -1 and cool down
```

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/AutoShardingType.java:
```diff
@@ -0,0 +1,16 @@
+
+package org.elasticsearch.action.datastreams.autosharding;
+
+public enum AutoShardingType {
```
Can you add docs for this class?

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,424 @@
+            Index oldestBackingIndex = dataStream.getIndices().get(0);
+            IndexMetadata oldestIndexMeta = metadata.getIndexSafe(oldestBackingIndex);
+            logger.debug(
+                "data stream autosharding service recommends no change in the number of shards for data stream [{}]  "
+                    + "with previous autosharding event [{}] and oldest generation index [{}] being created on [{}] ",
```
Should this log message include something about the reason being that the remaining reduce shards cooldown being `N` (which is > 0)?

andreidan replied:
This is not a thing anymore with the restructuring of the statuses.

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,424 @@
+                AutoShardingType.NO_CHANGE_REQUIRED,
+                writeIndex.getNumberOfShards(),
+                writeIndex.getNumberOfShards(),
+                TimeValue.ZERO,
```
It seems strange to me that we return 0 for the remaining cooldown, but we actually still have a cooldown time for the decreasing.

Do you think instead, maybe we should change the enum to have things like `COOLDOWN_PREVENTED_DECREASE` and `COOLDOWN_PREVENTED_INCREASE` to indicate this exactly, differently than `NO_CHANGE_REQUIRED`? (because we didn't actually compute whether we *would* have decreased in this case)

dakrone replied:
The nice thing about this is that then we don't have to rely on the caller checking the cooldown remaining for `INCREASE/DECREASE`, instead, they can do a `switch` statement on the enum value, and see either `INCREASE_SHARDS`, `DECREASE_SHARDS`, `NO_CHANGE_REQUESTED`, `NOT_APPLICABLE`, `COOLDOWN_PREVENTED_DECREASE`, or `COOLDOWN_PREVENTED_INCREASE` and make the appropriate logging/decision. The time remaining becomes more of an informative thing instead of a decision-making thing.

dakrone replied:
I also like this because then we could return "oh here's what the number of shards _WOULD BE_ if you increased/decreased right now", which can help us in the future debug whether we need to tune the cooldown periods.

andreidan replied:
I've added this suggestion and it's looking pretty neat. Thanks ! 

Love that the outside world doesn't have to care about the cooldown value. 

Note that now the `NO_CHANGE_REQUIRED` continues to have a cooldown of zero (which IMO makes sense) as otherwise one of the other statuses is returned.

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,424 @@
+     */
+    public static final Setting<TimeValue> DATA_STREAMS_AUTO_SHARDING_INCREASE_SHARDS_COOLDOWN = Setting.timeSetting(
+        "data_streams.auto_sharding.increase_shards.cooldown",
+        TimeValue.timeValueMinutes(15),
```
As we discussed earlier with @henningandersen, we could potentially lower this to be the same as the periodic polling for the data stream lifecycle service, which makes sense (so that during master failure we wait for at least N (default 5) minutes before scaling up shards.

The other question I have is whether we should just go ahead and make it a factor (still configurable) of the poll interval, so we could do something like `1.0` or `.9` and make it 90% of 5 minutes (so 270 seconds instead of 300).

I don't think we have to figure the relative part out in this PR necessarily (but I'd like Henning to comment on whether 5m is fine, since that's what we discussed earlier).

andreidan replied:
++ note that my understanding was that we'll be configuring it to slightly lower than the DSL poll interval - i.e. 4m - to make sure we **do** auto shard on subsequent data stream lifecycle runs (barring master failovers or shutdowns)

andreidan replied:
I've pushed 270 seconds as part of this PR but happy to amend it as needed

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,424 @@
+
+        return dataStream.getAutoShardingEvent() == null
+            // without a pre-existing auto sharding event we wait until the oldest index has been created longer than the decrease_shards
+            // cool down period "ago" so we don't immediately after a data stream is created, reduce the number of shards
```
Super minor, but missing a word between "we don't" and "immediately" I think, did you mean something like
```suggestion
            // cool down period "ago" so we don't immediately reduce the number of shards after a data stream is created
```
?

dakrone commented on server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java:
```diff
@@ -0,0 +1,424 @@
+            AutoShardingType.NO_CHANGE_REQUIRED,
+            writeIndex.getNumberOfShards(),
+            writeIndex.getNumberOfShards(),
+            TimeValue.ZERO,
```
Same comment here about the weirdness of passing 0 when we could maybe expand the enum and give more info (including a calculation)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAutoShardingEvent.java:
```diff
@@ -0,0 +1,88 @@
+        SimpleDiffable<DataStreamAutoShardingEvent>,
+        ToXContentFragment {
+
+    public static final ParseField TRIGGER_DATA_STREAM_GENERATION = new ParseField("trigger_generation");
```
Do you think we need the generation? It seems a little low-level for something like this, and like something that may confuse a user; what do you think?

andreidan replied:
> If we never have a situation where the xcontent pieces get written out to state, then it's fine not to have one. If we do, then we'll need one to ensure we don't have an unrecoverable cluster due to serialization issues

For some reason I can't comment on this suggestion, but we are writting the auto shard event in the data stream metadata and it is tested as part of the `DataStreamTests` which is an `AbstractXContentSerializingTestCase`

> Do you think we need the generation? It seems a little low-level for something like this, and like something that may confuse a user; what do you think?

I thought it might be useful for potential debugging but happy to leave it out until we have an use case for it if you think it's too low level (note that it's not going to be displayed anywhere except the `GET _data_stream` API)

dakrone replied:
Let's leave the generation out for now, I think it'll be more confusing to the user.

andreidan replied:
Oh, it just dawned on me (sorry it's been a while so forgot about it) why I thought initially it might be useful. 

The modify data stream API can remove the index that triggered an autosharding event, and then re-add it to the data stream, except after returning it to the data stream it'll be on the last position in the data stream. Without the generation information (which is monotonically increasing) the output of the `GET _data_stream` API might be confusing as a very "old" index triggered auto sharding, however none of the new following indices have the auto sharding recommendation as the number of shards. 

Admittedly a rather peculiar use case I think the generation is useful in this case. 

What do you think?

dakrone replied:
I think it's okay to have that special case for now, I still thing generation is a number that will be confusing to a user because even though it's increasing, it isn't a one-to-one kind of thing. We can always add it at a later time if we think it would be helpful then.

andreidan replied:
Right, removing it now.

dakrone reviewed (commented):
I left more comments, but nothing really major, thanks for working on this Andrei! (and sorry it took so long to review). I think @henningandersen should weigh in on the scale-up cooldown just to make sure we understood him correctly — my understanding was that we wanted to lower it to be the same as the data lifecycle polling interval, is that correct?

andreidan committed:
Update server/src/main/java/org/elasticsearch/action/datastreams/autosharding/AutoShardingResult.java

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service

andreidan committed:
Introduce COOLDOWN_PREVENTED_INCREASE/DECREASE statuses and refactoring

andreidan committed:
Test and comment

andreidan committed:
Remove generation from the shard event

dakrone reviewed (approved):
LGTM, thanks for iterating on this Andrei! I left three really minor things.

andreidan committed:
Logging, humanreadable field

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-autosharding-service
